tx-compiler/src/compiler.rs

```diff
@@ -0,0 +1,299 @@
+
+        // Create the transaction script [CodeBlock].  If no tx_script is provided we use a
+        // Noop span block.
+        let tx_script = match tx_script {
+            Some(tx_script) => self
+                .assembler
+                .compile_in_context(tx_script, &mut assembly_context)
+                .map_err(TransactionError::CompileTxScriptFailed)?,
+            None => CodeBlock::new_span(vec![Operation::Noop]),
+        };
+
+        // Merge transaction script code block and epilogue code block
+        let tx_script_and_epilogue =
+            CodeBlock::new_join([tx_script.clone(), self.epilogue.clone()]);
+
+        // Merge prologue and note script tree
+        let prologue_and_notes = CodeBlock::new_join([self.prologue.clone(), note_program_tree]);
+
+        // Merge prologue, note tree, tx script and epilogue
+        let program_root = CodeBlock::new_join([prologue_and_notes, tx_script_and_epilogue]);
+
+        // Create [CodeBlockTable] from [AssemblyContext]
+        let cb_table = self
+            .assembler
+            .build_cb_table(assembly_context)
+            .map_err(TransactionError::BuildCodeBlockTableFailed)?;
+
+        // Create transaction program
+        let program = Program::with_kernel(program_root, self.tx_kernel.clone(), cb_table);
+
+        // Create compiled transaction
+        // TODO: we always have a digest for a transaction script (noop code block root if no user
+        // tx scrip is provided) - should we we change the [CompiledTransaction] struct to take a
+        // digest instead of a Option<Digest> fro the tx script.
+        Ok(CompiledTransaction::new(account_id, notes, Some(tx_script.hash()), program))
+    }
+
+    /// Returns a [CodeBlock] which contains the note program tree.
+    fn create_note_program_tree(
+        &mut self,
+        account_id: &AccountId,
+        notes: &[Note],
+        assembly_context: &mut AssemblyContext,
+    ) -> Result<CodeBlock, TransactionError> {
+        // Fetch target account interface
+        let target_account_procs = self
+            .account_procedures
+            .get(account_id)
+            .ok_or(TransactionError::AccountInterfaceNotFound(*account_id))?;
+
+        // Create note programs
+        let mut note_program_tree = notes
+            .iter()
+            .map(|note| {
+                let note_root = self
+                    .assembler
+                    .compile_in_context(note.script().code().clone(), assembly_context)
+                    .map_err(|_| TransactionError::CompileNoteScriptFailed)?;
+                let note_branches = collect_call_branches(&note_root);
+                verify_note_account_compatibility(
+                    note_root.hash(),
+                    &note_branches,
+                    target_account_procs,
+                )?;
+                Ok(CodeBlock::new_join([self.note_setup.clone(), note_root]))
```

Comment:
We want to do something different here:

1. Instead of joining note setup script with `note_root`, we should be joining it with a new `Call` block which calls the note script. It should should be something like `Join(note_setup_script, Call(note_root))`.
2. The actual `CodeBlock` for the note should be added to the unified call block table.

We might need to refactor this to make several passes over the notes:
- First pass: compile notes to get their `CodeBlock`'s and verify compatibility with accounts. At the end of this pass we'd have everything compiled and then we can convert `AssemblyContext` into a call block table.
- Second pass: join notes with note setup scripts and also insert note code blocks into the code block table built in the previous step.
- Third pass: join code blocks into a tree and combine them with prologue/epilogue etc.